File: TripKit/Classes/core/TransportKit/TKRoutePolyline.m
//
//  RouteAnnotation.m
//  Tracker
//
//  Created by Adrian Schönig on 13/04/10.
//  Copyright 2010 Adrian Schönig. All rights reserved.
//

#import "TKRoutePolyline.h"

#import <TripKit/TripKit-Swift.h>

@implementation TKRoutePolyline

+ (nullable instancetype)routePolylineForRoute:(id <TKDisplayableRoute>)originRoute {
  // create the array of coordinates for the line
  NSArray * pathPoints = [originRoute routePath];
  
  NSUInteger i, j, count = [pathPoints count];
  if (count == 0) {
    return nil;
  }
  
  CLLocationCoordinate2D coords[count];
  for (i = 0, j = 0; i < count && j < count; i += 1, j ++) {
    id obj = [pathPoints objectAtIndex:i];
    if ([obj respondsToSelector:@selector(coordinate)]) {
      coords[j] = [obj coordinate];
    } else {
      ZAssert(false, @"Bad input: Route %@ has bad path point: %@", originRoute, obj);
      return nil;
    }
  }
  coords[count - 1] = [[pathPoints objectAtIndex:count - 1] coordinate];
  
  TKRoutePolyline * routeAnnotation = (TKRoutePolyline *) [self polylineWithCoordinates:coords count:count];
  routeAnnotation.route = originRoute;
  return routeAnnotation;
}

+ (nullable MKGeodesicPolyline *)geodesicPolylineForAnnotations:(NSArray *)annotations
{
  if (annotations.count == 0) {
    return nil;
  }
  
  CLLocationCoordinate2D coordinates[annotations.count];
  for (NSUInteger i = 0; i < annotations.count; i++) {
    id<MKAnnotation> annotation = annotations[i];
    coordinates[i] = [annotation coordinate];
  }
  MKGeodesicPolyline *geodesicPolyline = [MKGeodesicPolyline polylineWithCoordinates:coordinates
                                                                               count:annotations.count];
  return geodesicPolyline;
}

@end

File: TripKit/Classes/core/TransportKit/TKRoutePolyline.swift
//
//  TKRoutePolyline.swift
//  TripKit
//
//  Created by Adrian Schoenig on 27/10/16.
//
//

import Foundation
import CoreLocation
import MapKit

// MARK: - TKDisplayableRoute, the input

@objc
public protocol TKDisplayableRoute {
  
  var routePath: [Any] { get } // objects that have a coordinate, e.g., <MKAnnotation> or CLLocation
  var routeColor: TKColor? { get }
  var routeDashPattern: [NSNumber]? { get }
  var routeIsTravelled: Bool { get }
  var selectionIdentifier: String? { get }
  
}

extension TKDisplayableRoute {
  @nonobjc public var routeColor: TKColor? { return nil }
  @nonobjc public var routeDashPattern: [NSNumber]? { return nil }
  @nonobjc public var routeIsTravelled: Bool { return true }
}

fileprivate class WrappedCoordinate {
  fileprivate let coordinate: CLLocationCoordinate2D
  init(_ coordinate: CLLocationCoordinate2D) {
    self.coordinate = coordinate
  }
}

extension MKGeodesicPolyline : TKDisplayableRoute {
  public var routePath: [Any] {
    var coordinates = [CLLocationCoordinate2D]()
    coordinates.reserveCapacity(pointCount)
    getCoordinates(&coordinates, range: NSRange(location: 0, length: pointCount))
    return coordinates.map { WrappedCoordinate($0) }
  }
  
  public var routeColor: TKColor? {
    return nil
  }
  
  public var routeDashPattern: [NSNumber]? {
    return nil
  }
  
  public var routeIsTravelled: Bool {
    return true
  }
  
  public var selectionIdentifier: String? {
    return nil
  }
  
}

public class TKRoutePolyline : MKPolyline {
  public var route: TKDisplayableRoute!
  
  public var selectionIdentifier: String?
  
  public convenience init?(route: TKDisplayableRoute) {
    let pathpoints = route.routePath
    guard !pathpoints.isEmpty else { return nil }
    
    let coordinates = pathpoints.compactMap { obj -> CLLocationCoordinate2D? in
      switch obj {
      case let coordinate as CLLocationCoordinate2D: return coordinate
      case let location as CLLocation: return location.coordinate
      case let annotation as MKAnnotation: return annotation.coordinate
      default: return nil
      }
    }
    
    self.init(coordinates: coordinates, count: coordinates.count)

    self.route = route
  }
  
  
  /// - Returns: A geodesic polyline connecting the annotations
  public static func geodesicPolyline(annotations: [MKAnnotation]) -> MKGeodesicPolyline? {
    guard !annotations.isEmpty else { return nil }
    
    let coordinates = annotations.map(\.coordinate)
    return MKGeodesicPolyline(coordinates: coordinates, count: coordinates.count)
  }
}
